Introduce CI for AWS - part 1

wainersm opened the pull request:
This is just part 1 of a series of commits to run the e2e tests nightly for AWS too.

The way it's now, tests are executed but some fails. In particular, the TestAwsCreatePeerPodWithLargeImage fails in such as bad fashion that the job gets cancelled. The good news is that the simple pod tests at least pass. Here is an execution on my fork: https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/13122445400

Below is a list of things I still have to work on to make it acceptable running alongside our CI (at this point, the job will skip because I won't configure the AWS credentials on this repo yet). Nevertheless, I'd like to have this part merged because there are other aws-unrelated changes I plan to submit to the workflows and I want to avoid keep rebasing & resolving conflicts in my fork.

What's next:

- backup code to delete the created resources on AWS because there are some occasions where the deletion  code of the e2e framework doesn't run, for example, the failure of the TestAwsCreatePeerPodWithLargeImage I mentioned below causes that problem.
- deal with the failing tests. Either disable or fix them.
- run with CRI-O
- make it more resilient. For example, sometimes the VPC is created on an Availability Zone where the default podvm instance type isn't available, so it fails all tests
- [updated] add a debug step
- [updated] make it work with mkosi podvm images
- [updated ] move common code to scripts

stevenhorsman commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -0,0 +1,173 @@
+          export TEST_PODVM_IMAGE="${{ env.PODVM_QCOW2 }}"
+          export TEST_E2E_TIMEOUT="90m"
+
+          make test-e2e
```
Can we get any debug logs in case things go wrong?

wainersm replied:
Good idea, I overlooked it completely. In a next PR hopefully

wainersm commented on .github/workflows/e2e_run_all.yaml:
```diff
@@ -174,6 +174,28 @@ jobs:
+    name: aws
+    if: |
+      github.event_name == 'workflow_dispatch'
+    needs: [podvm, image, prep_install]
```
oh, I forgot to say on the description...I couldn't make it work with mkosi and it's in my list to debug. The workflow supports both packer and mkosi image though.

bpradipt replied:
AWS works perfectly fine with mkosi. The default images made available (and part of instructions in confidentialcontainers.org) starting 0.11.0 are mkosi based.
Once you have created the mkosi raw image, use the `raw-to-ami.sh` script to upload and create AMI or you can use uplosi.

wainersm replied:
hmmm I chained these changes with the existing workflow that builds the mkosi-based image, but all AWS e2e tests failed. Maybe the problem was/is somewhere else. I will revisit that topic soon.

wainersm commented:
Hi @stevenhorsman !

> Not necessarily for this PR, but I'm wondering if we can separate some of the steps into scripts given that lots of this is duplicated with the other providers. I know we've discussed it before, but I can't recall if we deliberately rejected it, or not

IMO, we can and must separate into scripts. Share common code and avoid the nasty pull_request_target limitation. I can give it a try on a following PR.

wainersm commented:
Updated with fixes to workflow lint errors.

ldoktor reviewed (commented):
Hello Wainer, overall it looks quite good, I'd suggest tagging the resources using the workflow-id to allow an extra cleanup job that would iterate over resources and deleted the ones where workflow finished. To further simplify that, if possible, I'd like to add a `ci=true` tag to every resource that is generated. I guess currently that is not supported by peer-pods, right? I think it'd be a useful feature even there (for peer-pods VMs so people can add custom tags).

One thing I noticed is that the job seems to attempt to remove even VPCs it had not created. Would be nice to double-check and only cleanup when we created them... (not 100% sure about that)

Looking for the part 2 ;-)

wainersm commented:
Hi @ldoktor !

> Hello Wainer, overall it looks quite good, I'd suggest tagging the resources using the workflow-id to allow an extra cleanup job that would iterate over resources and deleted the ones where workflow finished. To further simplify that, if possible, I'd like to add a `ci=true` tag to every resource that is generated. I guess currently that is not supported by peer-pods, right? I think it'd be a useful feature even there (for peer-pods VMs so people can add custom tags).
> 

In part 2 that I should send soon I added an extra step that runs "always" to do such as cleanup in case the e2e finish abruptly. I just need to ensure it deletes only the resources created a that given workflow execution.

I never tried myself, but it should be possible to configure peer-pods to tag the podvms: https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/install/overlays/aws/kustomization.yaml#L35

> One thing I noticed is that the job seems to attempt to remove even VPCs it had not created. Would be nice to double-check and only cleanup when we created them... (not 100% sure about that)

It should delete only the VPCs created (and related resources) created by the e2e test run.

> 
> Looking forward for the part 2 ;-)

\o/

wainersm commented:
Rebased and resolved a little conflict in go.mod; also fixed some linter errors.

@bpradipt , my friend, can I have your ack?

@stevenhorsman does your ack stand?

bpradipt reviewed (approved):
/lgtm
Thanks @wainersm .. Sorry for the delay.

stevenhorsman commented:
> @stevenhorsman does your ack stand?

Yeah, I re-reviewed and it looks good. Thanks!
